### PR TITLE
Set tier on upload/copy; Rehydrate Priority

### PR DESCRIPTION
**Pull request by jaschrep-msft**

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 1]
```diff
@@ -254,11 +256,16 @@ public URL getBlobUrl() {
      * @param destAccessConditions {@link BlobAccessConditions} against the destination.
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: Need to add Javadocs for the parameters

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 2]
```diff
@@ -357,11 +364,15 @@ public URL getBlobUrl() {
...
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: Need to add Javadocs for the new parameters

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java [hunk 3]
```diff
@@ -145,18 +147,18 @@
      * @return A reactive response containing the information of the uploaded block blob.
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java]: Need to add Javadocs for the new parameters

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java [hunk 4]
```diff
@@ -239,7 +241,7 @@
...
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java]: Need to add Javadocs for the new parameters

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java [hunk 5]
```diff
@@ -325,7 +327,7 @@ as we can guarantee we only need at most two buffers for any call to write (two
      * @throws UncheckedIOException If an I/O error occurs
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java]: Need to add Javadocs for the new parameters

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java [hunk 6]
```diff
@@ -563,17 +567,21 @@ private String getBlockID() {
      * @return A reactive response containing the information of the block blob.
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlockBlobAsyncClient.java]: Need to add Javadocs for the new parameters

**Comment by jaschrep-msft** [comment]
Uploaded to fork as a development checkpoint. Clicked pull request on autopilot. Work up to this point is incomplete.

**Comment by alzimmermsft** [comment]
This fixes #4512

**Comment by xiafu-msft** [comment]
Based on the discussion offline, we want to add tests for rehydrate priority

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy [hunk 7]
```diff
@@ -1835,12 +1835,31 @@ class BlobAPITest extends APISpec {
...
+        setup:
+        def bc = cc.getBlockBlobClient(generateBlobName())
+        bc.uploadWithResponse(defaultInputStream.get(), defaultDataSize, null, null, tier1, null, null, null)
+        def bcCopy = cc.getBlockBlobClient(generateBlobName())
+
+        when:
+        bcCopy.copyFromURLWithResponse(new URL(bc.getBlobUrl().toString() + "?" + bc.generateSAS(OffsetDateTime.now().plusHours(1), new BlobSASPermission().read(true))), null, tier2, null, null, null, null)
+
+        then:
+        bcCopy.getPropertiesWithResponse(null, null, null).headers().value("x-ms-access-tier") == tier2.toString()
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: >.headers().value("x-ms-access-tier")  [](start = 58, length = 37)

I think this is already stored in BlobProperties.accessTier

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobAPITest.groovy [hunk 8]
```diff
@@ -739,14 +740,25 @@ class BlockBlobAPITest extends APISpec {
...
 
+    def "Upload with tier"() {
+        setup:
+        def bc = cc.getBlockBlobClient(generateBlobName())
+
+        when:
+        bc.uploadWithResponse(defaultInputStream.get(), defaultDataSize, null, null, AccessTier.COOL, null, null, null)
+
+        then:
+        AccessTier.fromString(bc.getPropertiesWithResponse(null, null, null).headers().value("x-ms-access-tier")) == AccessTier.COOL
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobAPITest.groovy]: >.headers().value("x-ms-access-tier")) [](start = 76, length = 37)

Same as other comment

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy (hunk 7 above)
Comment (jaschrep-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: It is not.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobAPITest.groovy (hunk 8 above)
Comment (jaschrep-msft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlockBlobAPITest.groovy]: Same as previous.

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy (hunk 7 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/BlobAPITest.groovy]: Fixes this in #5268, will clean this up when I merge these changes into that branch

**Pull request merged**